### PR TITLE
Add NSys Profiling Interface to NeMo Core

### PR DESCRIPTION
**Pull request by erhoo82**
# What does this PR do ?

The interface to generate nsys profile results. 

**Collection**: [Core]

# Changelog 
- Enable profile
- Specify range to profile by batch step
- Specify rank to profile
- Specify profiling details

# Usage
Enable knobs to profile

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 1]
```diff
@@ -273,6 +284,15 @@ def training_step(self, batch, batch_idx):
 
         return loss_mean
 
+
+    def on_train_batch_start(self, batch, batch_idx, unused: Optional[int] = 0) -> None:
+        if batch_idx == self._profile_start_step and torch.distributed.get_rank() == self._profile_rank:
+            print("====== Start nsys profiling ======")
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:290]: Can we change these to `logging.info()` messages?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -311,6 +331,10 @@ def on_train_batch_end(self, outputs, batch, batch_idx: int, unused: Optional[in
                     # accumulated gradient updates.
                     grad_scaler.optimizer_update_skipped = None
 
+        if batch_idx == self._profile_end_step and torch.distributed.get_rank() == self._profile_rank:
+            print("====== End nsys profiling ======")
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:335]: Same here, can we make these `logging.info()` calls?

**Review by MaximumEntropy — COMMENTED** [review]
Minor comments about logging.

File: examples/nlp/language_modeling/conf/megatron_gpt_config.yaml [hunk 3]
```diff
@@ -113,6 +113,13 @@ model:
   # See Reducing Activation Recomputation in Large Transformer Models: https://arxiv.org/abs/2205.05198 for more details.
   sequence_parallel: False
 
+  ## Nsys profiling options
+  nsys_profile: False
```
Comment (titu1994) [inline examples/nlp/language_modeling/conf/megatron_gpt_config.yaml:117]: Can you make it a nested config. IE

```yaml
nsys_profiler:
  profile: False
  profile_start_step: 10  # Global batch to start profiling
  profile_end_step: 10 # Global batch to end profiling
  profile_ranks: [0] # Global rank IDs to profile
  profile_gen_shape: False # Generate model and kernel details including input shapes

```

File: nemo/core/classes/modelPT.py [hunk 4]
```diff
@@ -1400,3 +1403,60 @@ def update_save_restore_connector(cls, save_restore_connector):
...
         else:
             setattr(cls, '_save_restore_connector', save_restore_connector)
+
+    def _setup_nsys_profiling(self):
+        """ Enables nsys profiling
+
+            To use, add the following optoins to the model config:
+
+            ## Nsys profiling options
+            nsys_profile: False
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1413]: Update syntax for nested config

File: nemo/core/classes/modelPT.py [hunk 5]
```diff
@@ -1400,3 +1403,60 @@ def update_save_restore_connector(cls, save_restore_connector):
...
+
+            And then wrap the model training script with:
+            nsys profile -s none -o <profile filepath>  -t cuda,nvtx --force-overwrite true --capture-range=cudaProfilerApi --capture-range-end=stop python ./examples/...
+
+            See more options at: https://docs.nvidia.com/nsight-systems/UserGuide/index.html#cli-profiling
+
+        """
+        if self.cfg.get('nsys_profile', False):
+            # Nsys profiling options
+            self._nsys_profile_start_step = self.cfg.nsys_profile_start_step
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1427]: These values should be obtained via self.cfg.get(xyz, <reasonable default>)

**Review by titu1994 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**